### PR TITLE
Handle complex URL [e.g. single video AND playlist in the same YouTube URL; e.g. YouTube URLs containing @channelname]

### PR DESCRIPTION
**Pull request by deldesir**
🚀 Pull Request Overview:

Any URL like https://www.youtube.com/watch?v=phWxA89Dy94&list=PLuw-0UPDDxLYTaK9_36SepO6leuLC5_pJ&pp=iAQB will fail on xklb, then gets its second query parameter stripped before lb tubeadd runs again.

This PR removes the secondary query parameter to avoid this failure. It also handles channels and URLs like https://www.youtube.com/@blondesiro3209/playlists.

📋 Checklist:

 - [x] Tested the change on 10.8.0.22 (Ubuntu 24.04) with URL https://www.youtube.com/watch?v=phWxA89Dy94&list=PLuw-0UPDDxLYTaK9_36SepO6leuLC5_pJ&pp=iAQB
 
**Without the PR (multiple times):**
 
failed

**With the PR applied (twice):**

success

Related issue: https://github.com/iiab/calibre-web/issues/104

cc @EMG70

### CODE REVIEW

File: cps/tasks/metadata_extract.py [hunk 1]
```diff
@@ -86,22 +92,26 @@ def run(self, worker_thread):
                         log.error("An error occurred while trying to connect to the database: %s", db_error)
                         self.message = f"{self.media_url_link} failed: {db_error}"
 
-                    # get the shelf title
```
[the remark this answers is not in the document]
Comment (deldesir) [reply cps/tasks/metadata_extract.py:89]: You're right. I added it back.

**Comment by holta** [comment]
> This PR removes the secondary query parameter to avoid this failure.

If the leftmost parameter is kept and all others deleted, this should be documented (1) for devs in-line, and also for (2) users at https://github.com/iiab/calibre-web/wiki

> It also handles channels and URLs like https://www.youtube.com/@blondesiro3209/playlists

How?

Thanks for explaining!

**Comment by holta** [comment]
Can screenshot(s) illustrate what happens with URLs like?

- https://youtube.com/@blondesiro3209
- https://youtube.com/@blondesiro3209/playlists

**Comment by EMG70** [comment]
IIAB-DIAGNOSTICS http://sprunge.us/CTP9XU?en
Before applying PR#134, Download of https://www.youtube.com/watch?v=phWxA89Dy94&list=PLuw-0UPDDxLYTaK9_36SepO6leuLC5_pJ&pp=iAQB  resulted in same errors as shown on @deldesir screenshot above.
After applying PR#134, Previously failed videos and playlists downloaded OK
Screenshot from 2024-02-21 15-09-40

**Comment by holta** [comment]
> failed to download: 'NoneType' object is not subscriptable

@deldesir can you explain both above errors?

@EMG70 huge thanks for your screenshots!

**Comment by EMG70** [comment]
Bear with me whilst i have a clean VM as the failed long playlists are now being redownloaded when testing PR#131

**Comment by EMG70** [comment]
> Can screenshot(s) illustrate what happens with URLs like?
> 
>     * https://youtube.com/@blondesiro3209
> 
>     * https://youtube.com/@blondesiro3209/playlists
IIAB-DIAGNOSTICS   http://sprunge.us/EuCAYH?en

Another download of  https://youtube.com/@blondesiro3209/playlists was done after a fresh VM was created.All videos still  downloading ok ,so far 24/383 videos successfully downloaded.
Screenshot from 2024-02-21 18-29-11
Screenshot from 2024-02-21 18-35-31

**Comment by deldesir** [comment]
Following videos failed for some reason I don't know yet. 

https://m.youtube.com/watch?v=ZfbiLfSczgA

https://m.youtube.com/watch?v=Zx0gMxl5qjY

Generally when the video is playable and Type error is returned, I assume the requested format is not found.

**Comment by EMG70** [comment]
At the time of me wrapping up for the day,the following Blondle's playlist of 383 videos had downloaded 13+100+114+7=234.That's 234/383.
Screenshot from 2024-02-21 18-29-11

Screenshot from 2024-02-21 21-02-12

**Comment by EMG70** [comment]
IIAB-DIAGNOSTICS http://sprunge.us/UwhHce?en

You will find failed videos in iiab-diagnostics towards the end,this was due to low space on disk VM's disk.see screenshots below.
Screenshot from 2024-02-21 20-48-41
Screenshot from 2024-02-21 21-08-22

**Comment by holta** [comment]
> @holta I have yet to find a short channel to provide a realistic complete test screenshot.

@EMG70 might recommend something?!

**Comment by holta** [comment]
> Following videos failed for some reason I don't know yet.
> 
> https://m.youtube.com/watch?v=ZfbiLfSczgA
> 
> https://m.youtube.com/watch?v=Zx0gMxl5qjY
> 
> Generally when the video is playable and Type error is returned, I assume the requested format is not found.

Reposted to new ticket:

- #135

Please investigate and follow up there.

**Comment by EMG70** [comment]
> > @holta I have yet to find a short channel to provide a realistic complete test screenshot.
> 
> @EMG70 might recommend something?!

This channel https://www.youtube.com/@idata.analyticshub has got 26 videos,they all downloaded Ok.
IIAB-DIAGNOSTICS - http://sprunge.us/aYsfrz?en
Screenshot from 2024-02-22 15-02-57
Screenshot from 2024-02-22 15-03-14

**Comment by holta** [comment]
@deldesir will try to modify `scripts/lb-wrapper` to force fragmented video downloads (like #135!) to abort.

This should hopefully solve the freezing problem!

**Comment by deldesir** [comment]
> @deldesir will try to modify `scripts/lb-wrapper` to force fragmented video downloads (like #135!) to abort.
> 
> This should hopefully solve the freezing problem!

Done in https://github.com/deldesir/calibre-web/commit/4def7d1bc9b86459eec8c35a781b3c26ddb1579e

**Comment by holta** [comment]
> Done in [deldesir@4def7d1](https://github.com/deldesir/calibre-web/commit/4def7d1bc9b86459eec8c35a781b3c26ddb1579e)

Reducing retries from yt-dlp default 10 (or xklb default 12) to 1 will cause other video downloads to fail, e.g. when user's ISP is overloaded:

- Damaging overall reliability is not the way forward 😞 

- Instead: Consider conditionally acting on yt-dlp's exit code (error code) to determine which video downloads failed to download — despite 10-or-12 retries — at which point mark them as "not worth downloading" etc.

**Comment by holta** [comment]
Degrading all yt-dlp downloads by making video downloads fragile (e.g. avoiding retries) would be a step backward — not a step forward — for users needing a very reliable & predictable tool.

Coherent Error Handling is what's needed:

1) Accept that yt-dlp failing to download particular videos is inevitable.

2) No matter if due to yt-dlp bugs like #135 — due to ISP slowness / overloading / overselling bandwidth / ISP interruptions — and/or for any other reasons that will emerge later.

3) These failures then need to be (1) auto-identified and (2) programmatically acted upon.  e.g. by marking those videos as "not worth downloading" or "failed to download video" or similar.

**Comment by holta** [comment]
@deldesir please explain or summarize how this PR #134 should evolve.

Either based on...

- PR chapmanjacobd/library#32

...and/or based on any other pragmatic ideas that have come together over the past week — it's time to push this forward in one direction or another, Thanks! 🗝️

**Comment by holta** [comment]
Or if a separate PR is more appropriate, that is 100% fine too 🙂

**Comment by holta** [comment]
Very illuminating discussion began ~6 hours ago — both outlining why YouTube appears to be causing #135 yt-dlp downloads to fail — and also hinting at possible pragmatic &/or long-term solutions:

- https://github.com/yt-dlp/yt-dlp/issues/3372#issuecomment-1973750981

**Comment by holta** [comment]
@deldesir can you summarize on https://github.com/iiab/calibre-web/wiki which URL's are most recommended — and which URL's are most risky❔

(So it's very very to almost everyone, what to expect after they click "Download to IIAB".)

**Comment by deldesir** [comment]
Done. Please have a look at https://github.com/iiab/calibre-web/wiki/Home/_compare/32fa33e1d35d9e321c41e15decaf41c0e6cce1fe...8376933a218f164fa18f3cf378ce491501dd94e7
